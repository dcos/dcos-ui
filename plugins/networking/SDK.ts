let SDK;

module.exports = {
  getSDK() {
    return SDK;
  },
  setSDK(pluginSDK) {
    SDK = pluginSDK;
  }
};
