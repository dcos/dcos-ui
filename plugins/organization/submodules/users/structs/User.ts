import Account from "../../../structs/Account";

import GroupsList from "../../groups/structs/GroupsList";

export default class User extends Account {
  // NOTE: getGroups cannot be moved to Account because of circular
  // reference issues
  getGroups() {
    const groups = this.get("groups") || [];
    const items = groups.map((groupMembership) => groupMembership.group);

    return new GroupsList({ items });
  }

  isRemote() {
    return this.get("is_remote");
  }
}
