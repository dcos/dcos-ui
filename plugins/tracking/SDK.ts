let SDK;

export default {
  getSDK() {
    return SDK;
  },
  setSDK(pluginSDK) {
    SDK = pluginSDK;
  }
};
