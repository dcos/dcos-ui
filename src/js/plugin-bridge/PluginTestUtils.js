import JestUtil from '../utils/JestUtil';
import Loader from './Loader';
import PluginModules from './PluginModules';
import PluginSDK from 'PluginSDK';

/**
 * Loads whatever plugins are passed in. Could be Mocks
 * @param  {Object} plugins - Plugin name to Config
 */
function loadPlugins(plugins) {
  var availablePlugins = {};
  var pluginConfig = {};

  Object.keys(plugins).forEach(function (pluginID) {
    availablePlugins[pluginID] = plugins[pluginID].module;
    pluginConfig[pluginID] = plugins[pluginID].config;
  });

  Loader.__setMockPlugins(availablePlugins);
  PluginSDK.initialize(pluginConfig);
}

/**
 * Finds actual plugins by name and loads them
 * @param  {Object} plugins - Map of name to config
 */
function loadPluginsByName(plugins) {
  let pluginsToLoad = {};

  let {pluginsList, externalPluginsList} = Loader.__getAvailablePlugins();

  Object.keys(plugins).forEach(function (pluginID) {

    if (!(pluginID in pluginsList) && !(pluginID in externalPluginsList)) {
      throw new Error(`${pluginID} does not exist. Failed to load.`);
    }
    let path;
    // Default to always selecting internal plugin if same pluginID
    // exists in external plugins. This makes mocking easier.
    if (pluginID in pluginsList) {
      path = pluginsList[pluginID];
    } else {
      path = externalPluginsList[pluginID];
    }
    pluginsToLoad[pluginID] = {
      module: path,
      config: plugins[pluginID]
    };
  });
  loadPlugins(pluginsToLoad);

}

/**
 * Finds and returns the SDK for a specific plugin
 * @param  {String} pluginID - ID of plugin
 * @param  {Object} config   - configuration
 * @param  {Boolean} loadPlugin
 * @return {PluginSDK}          - SDK for plugin with pluginID
 */
function getSDK(pluginID, config, loadPlugin = false) {
  // Load plugin first so it can register it's reducer
  // if it has one.
  if (loadPlugin) {
    loadPluginsByName({
      [pluginID]: config
    });
  }
  // Get SDK for pluginID.
  return PluginSDK.__getSDK(pluginID, config);
}

/**
 * Set a mock for a module which would normally be returned by
 * PluginSDK.get().
 * @param {String} name - name of module
 * @param {any} mock - value representing the mock
 * @returns {mock} - the mock that was passed in
 */
function setMock(name, mock) {
  Loader.__setMockModule(name, mock);
  return mock;
}

/**
 * Add reducer to Store. Could be an actual plugin reducer or a mock for test
 * cases that require specific state.
 * @param {String} root    - Root key for which reducer will manage state
 * @param {Function} reducer - Reducer function
 */
function addReducer(root, reducer) {
  PluginSDK.__addReducer(root, reducer);
}

/**
 * Takes an Array or String representing a module name{s) that need unMocking,
 * finds the directory of the file and calls jest.dontMock on path.
 * @param  {Array|String} moduleNames - Names of modules to dontMock

 */
function dontMock(moduleNames) {
  if (Array.isArray(moduleNames)) {
    moduleNames.forEach(dontMock);
    return;
  }
  // Just one module to mock
  var name = moduleNames;
  // Try unmocking store first
  if (JestUtil.dontMockStore(name)) {
    return;
  }
  // Assuming modules have unique names
  let foundType = Object.keys(PluginModules).filter(moduleType => {
    return name in PluginModules[moduleType];
  });
  if (!foundType.length) {
    throw new Error(`Module ${name} does not exist.`);
  }
  let modulePath = `../${foundType[0]}/${PluginModules[foundType[0]][name]}`;
  jest.dontMock(modulePath);
}

let TestUtils = {
  addReducer,
  getSDK,
  dontMock,
  loadPlugins,
  loadPluginsByName,
  setMock
};

module.exports = TestUtils;
