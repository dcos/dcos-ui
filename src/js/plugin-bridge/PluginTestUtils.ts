import PluginSDK from "PluginSDK";

/**
 * Finds and returns the SDK for a specific plugin
 * @param  {String} pluginID - ID of plugin
 * @param  {Object} config   - configuration
 * @return {PluginSDK}          - SDK for plugin with pluginID
 */
function getSDK(pluginID, config) {
  // Get SDK for pluginID.
  return PluginSDK.__getSDK(pluginID, config);
}

/**
 * Add reducer to Store. Could be an actual plugin reducer or a mock for test
 * cases that require specific state.
 * @param {String} root    - Root key for which reducer will manage state
 * @param {Function} reducer - Reducer function
 */
function addReducer(root, reducer) {
  PluginSDK.__addReducer(root, reducer);
}

const TestUtils = {
  addReducer,
  getSDK
};

export default TestUtils;
