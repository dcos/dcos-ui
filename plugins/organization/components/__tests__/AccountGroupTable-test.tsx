import * as React from "react";
import { shallow } from "enzyme";

import User from "../../submodules/users/structs/User";

import PluginTestUtils from "PluginTestUtils";

const SDK = PluginTestUtils.getSDK("organization", { enabled: true });
require("../../SDK").setSDK(SDK);

const ACLGroupStore = require("../../submodules/groups/stores/ACLGroupStore")
  .default;
const ACLUserStore = require("../../submodules/users/stores/ACLUserStore")
  .default;
const AccountGroupTable = require("../AccountGroupTable").default;
const ActionTypes = require("../../submodules/groups/constants/ActionTypes");

const userDetailsFixture = require("../../../../tests/_fixtures/acl/user-with-details.json");

userDetailsFixture.groups = userDetailsFixture.groups.array;

let thisUserStoreGetUser, thisInstance;

describe("AccountGroupTable", () => {
  beforeEach(() => {
    thisUserStoreGetUser = ACLUserStore.getUser;

    ACLUserStore.getUser = userID => {
      if (userID === "unicode") {
        return new User(userDetailsFixture);
      }
    };
    const getAccountDetails = () => ACLUserStore.getUser("unicode");

    thisInstance = shallow(
      <AccountGroupTable
        accountID={"unicode"}
        getAccountDetails={getAccountDetails}
      />
    );

    thisInstance.instance().handleOpenConfirm = jest.fn();
  });

  afterEach(() => {
    ACLUserStore.removeAllListeners();
    ACLGroupStore.removeAllListeners();
    ACLUserStore.getUser = thisUserStoreGetUser;
  });

  describe("#onAclGroupStoreDeleteUserError", () => {
    it("updates state when an error event is emitted", () => {
      ACLGroupStore.deleteUser = () => {
        SDK.dispatch({
          type: ActionTypes.REQUEST_ACL_GROUP_REMOVE_USER_ERROR,
          data: "foo bar",
          groupID: "baz",
          userID: "unicode"
        });
      };

      ACLGroupStore.deleteUser("foo", "unicode");
      expect(thisInstance.state("userUpdateError")).toEqual("foo bar");
    });
  });

  describe("#onAclGroupStoreDeleteUserSuccess", () => {
    it("gets called when a success event is emitted", () => {
      ACLGroupStore.deleteUser = () => {
        SDK.dispatch({
          type: ActionTypes.REQUEST_ACL_GROUP_REMOVE_USER_SUCCESS,
          data: "foo bar",
          groupID: "baz",
          userID: "unicode"
        });
      };
      thisInstance.instance().onAclGroupStoreDeleteUserSuccess = jest.fn();

      ACLGroupStore.deleteUser("foo", "unicode");
      expect(
        thisInstance.instance().onAclGroupStoreDeleteUserSuccess.mock.calls
          .length
      ).toEqual(1);
    });
  });

  describe("#getConfirmModalContent", () => {
    beforeEach(() => {
      thisInstance.setState({ groupID: "bar" });
    });

    it("returns a message with the user's name and group name", () => {
      const modalContent = thisInstance.instance().getConfirmModalContent({
        uid: "zed",
        description: "foo",
        groups: [{ group: { gid: "bar", description: "qux" } }]
      });

      const component = shallow(modalContent);
      expect(component).toMatchSnapshot();
    });

    it("returns a message containing the error that was received", () => {
      thisInstance.setState({ userUpdateError: "quux" });
      const modalContent = thisInstance.instance().getConfirmModalContent({
        description: "foo",
        groups: [{ group: { gid: "bar", description: "qux" } }]
      });

      const component = shallow(modalContent);
      expect(
        component
          .find("p")
          .at(0)
          .text()
      ).toEqual("quux");
    });
  });

  describe("#renderGroupLabel", () => {
    it("returns the specified property from the object", () => {
      const label = thisInstance
        .instance()
        .renderGroupLabel("foo", { gid: 1, foo: "bar" });
      expect(label.props.to).toEqual("/organization/groups/1");
      expect(label.props.children).toEqual("bar");
    });
  });

  describe("#renderButton", () => {
    it("calls handleOpenConfirm with the proper arguments", () => {
      const buttonWrapper = shallow(
        thisInstance.instance().renderButton("foo", { gid: "bar" })
      );

      buttonWrapper.find("button").simulate("click");

      expect(
        thisInstance.instance().handleOpenConfirm.mock.calls[0][0]
      ).toEqual({
        gid: "bar"
      });
    });
  });
});
