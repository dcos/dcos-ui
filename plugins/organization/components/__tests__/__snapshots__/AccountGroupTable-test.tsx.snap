// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountGroupTable #getConfirmModalContent returns a message with the user's name and group name 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <div>
    <WithI18n
      render="p"
    >
      zed
       will be removed from 
      bar
      .
    </WithI18n>
  </div>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <WithI18n
          render="p"
        >
          zed
           will be removed from 
          bar
          .
        </WithI18n>,
        null,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            "zed",
            " will be removed from ",
            "bar",
            ".",
          ],
          "render": "p",
        },
        "ref": null,
        "rendered": Array [
          "zed",
          " will be removed from ",
          "bar",
          ".",
        ],
        "type": [Function],
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <WithI18n
            render="p"
          >
            zed
             will be removed from 
            bar
            .
          </WithI18n>,
          null,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "zed",
              " will be removed from ",
              "bar",
              ".",
            ],
            "render": "p",
          },
          "ref": null,
          "rendered": Array [
            "zed",
            " will be removed from ",
            "bar",
            ".",
          ],
          "type": [Function],
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Array [],
}
`;
